### ISSUE TITLE
excessive time on insert select on tables with large columns (~2000) even inserting a single row 

### ISSUE THREAD

**Issue by shawel** [ClickHouse/ClickHouse#12408]
i have a table with about 2000 columns and 1.7 billion rows.  I wanted to insert select to a different table with limited number of rows usually in the 1000s for dashboards. But the time inserting even a single row takes significant amount of time and memory explodes.  I have a single order by column device_xx64 which is a UINT64 and the rest of the 2000 columns are UInt8 arrays with an average row size of about 600 bytes. 

Here is a count of the insert select (super fast) 
```

SELECT count(device_xx64)
FROM device_table
PREWHERE device_xx64 IN 
(
    SELECT device_xx64
    FROM device_mem
    LIMIT 100
)
┌─count(device_xx64)─┐
│                100 │
└────────────────────┘

1 rows in set. Elapsed: 0.133 sec. Processed 19.89 thousand rows, 159.12 KB (149.42 thousand rows/s., 1.20 MB/s.) 
```

But inserting even 1 row takes significant amount of time and memory.

```

INSERT INTO test_1 SELECT 
    *,
    device_xx64
FROM device_table
PREWHERE device_xx64 IN 
(
    SELECT device_xx64
    FROM device_mem
    LIMIT 1
)

Ok.

0 rows in set. Elapsed: 139.411 sec. Processed 11.70 thousand rows, 12.04 MB (83.92 rows/s., 86.38 KB/s.) 
```

it gets faster on repeated inserts of same data to about 10 seconds. but that does not help me. 

Seems like a performance issues? If it cannot be fixed there a faster alternative? 

Thank you. 

I have tried on clickhouse version 20+

**Comment by den-crane** [comment]
Please provide test_1 DDL. 
It's expected that such insert is slow. Because it need to create 4000 files (bin+mrk).
But 139.411 sec is too slow
Also such insert will use 4GB memory (2MB*2000 (1MB compress buffer + 1MB write buffer))

**Comment by shawel** [comment]
I see! is there a way to speed it up? idk if it is slow on the write side or the read side. it seems to me like it it might be reading
to many unnecessary rows initially. just a hunch though.  My machine is 16 cores and 125 GB memory and 2TB EBS

This query climbs to using abour 30GB from 2 gb or so as the query goes on. 

both test_1 DDL and device_table have same ddl which is 

```
CREATE TABLE default.surf_audience
(
    `device_uuid` UUID CODEC(NONE), 
    `segment_size` SimpleAggregateFunction(sum, UInt64), 
    `device_xx64` UInt64 MATERIALIZED xxHash64(device_uuid), 
    `S1339392` SimpleAggregateFunction(max, Array(UInt8)) DEFAULT [], 
    `S1339136` SimpleAggregateFunction(max, Array(UInt8)) DEFAULT [], 
    `S1340160` SimpleAggregateFunction(max, Array(UInt8)) DEFAULT [], 
    .... 2000 more of the same SimpleAggregateFunction(max, Array(UInt8)) DEFAULT [], type
ENGINE = AggregatingMergeTree()
PARTITION BY tuple()
ORDER BY device_xx64
SETTINGS index_granularity = 8192
```

**Comment by filimonov** [comment]
May be related: https://github.com/ClickHouse/ClickHouse/issues/6943

**Comment by shawel** [comment]
also maybe related to this #7187 pehraps ???

**Comment by alexey-milovidov** [comment]
Fixed in version 20.10-prestable by enabling compact data parts.
